### ISSUE TITLE
Serve DLNA lists only images on Android

### ISSUE THREAD

**Issue by serovar** [rclone/rclone#3475]
#### What is the problem you are having with rclone?
If I start a DLNA server with rclone serve dlna myremote: on Android (with Termux), it only lists image files inside the folders, and I can't see any other filetype. I tried with a lot of DLNA clients with the same results, and using the same command from Linux (Fedora 64 bit) works fine.

#### What is your rclone version (output from `rclone version`)
1.48.0 and 1.48.0-DEV (I tried with the Termux rclone package, with the official one built from source and with the official one built from the fix-dlna branch, with the same results).

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Android 9 (Pie) 64 bit

####  Which cloud storage system are you using? (eg Google Drive)
Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
rclone serve dlna Drive-crypted:

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

<details>
<summary>Using the rclone package from termux or the official one built from source:</summary>
$ rclone serve -vv dlna Drive-crypted:
2019/08/22 18:30:37 DEBUG : rclone: Version "v1.48.0" starting with parameters ["rclone" "serve" "-vv" "dlna" "Drive-crypted:"]
2019/08/22 18:30:37 DEBUG : Using config file from "/data/data/com.termux/files/home/.config/rclone/rclone.conf"
2019/08/22 18:30:39 DEBUG : Adding path "vfs/forget" to remote control registry
2019/08/22 18:30:39 DEBUG : Adding path "vfs/refresh" to remote control registry
2019/08/22 18:30:39 DEBUG : Adding path "vfs/poll-interval" to remote control registry
2019/08/22 18:30:39 NOTICE: Encrypted drive 'Drive-crypted:': Serving HTTP on [::]:7879
2019/08/22 18:30:39 INFO  : DLNA server on :7879: Started SSDP on dummy0
2019/08/22 18:30:39 INFO  : DLNA server on :7879: Started SSDP on rmnet_ipa0
2019/08/22 18:30:39 INFO  : DLNA server on :7879: Started SSDP on lo
2019/08/22 18:30:39 INFO  : DLNA server on :7879: Started SSDP on rmnet_data0
2019/08/22 18:30:39 INFO  : DLNA server on :7879: Started SSDP on wlan0
2019/08/22 18:30:39 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:39 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:39 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:39 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:39 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:39 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:40 INFO  : /static/ContentDirectory.xml: 192.168.1.10:48688 GET 200  %!s(<nil>)
2019/08/22 18:30:40 INFO  : /static/ConnectionManager.xml: 192.168.1.10:48688 GET 200  %!s(<nil>)
2019/08/22 18:30:40 INFO  : /static/X_MS_MediaReceiverRegistrar.xml: 192.168.1.10:48688 GET 200  %!s(<nil>)
2019/08/22 18:30:49 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:49 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:49 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:49 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:49 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:49 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:53 INFO  : /static/ContentDirectory.xml: 192.168.1.8:38541 GET 200  %!s(<nil>)
2019/08/22 18:30:53 INFO  : /static/ConnectionManager.xml: 192.168.1.8:38541 GET 200  %!s(<nil>)
2019/08/22 18:30:53 INFO  : /static/X_MS_MediaReceiverRegistrar.xml: 192.168.1.8:38541 GET 200  %!s(<nil>)
2019/08/22 18:30:54 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.doc
2019/08/22 18:30:54 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.mp4
2019/08/22 18:30:54 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.rar
2019/08/22 18:30:54 INFO  : /ctl: 192.168.1.8:38541 POST 200 "urn:schemas-upnp-org:service:ContentDirectory:1#Browse" %!s(<nil>)
2019/08/22 18:30:54 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.doc
2019/08/22 18:30:54 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.mp4
2019/08/22 18:30:54 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.rar
2019/08/22 18:30:54 INFO  : /ctl: 192.168.1.8:38541 POST 200 "urn:schemas-upnp-org:service:ContentDirectory:1#Browse" %!s(<nil>)
2019/08/22 18:30:59 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:59 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:59 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:59 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:59 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:30:59 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:31:09 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:31:09 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:31:09 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:31:09 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:31:09 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted
2019/08/22 18:31:09 error writing to UDP socket: write udp 0.0.0.0:1900->239.255.255.250:1900: sendto: operation not permitted

</details>

<details>
<summary>Using the official one built from the fix-dlna branch:</summary>
$ rclone/rclone serve -vv dlna Drive-crypted:
2019/08/22 18:20:08 DEBUG : rclone: Version "v1.48.0-DEV" starting with parameters ["rclone/rclone" "serve" "-vv" "dlna" "Drive-crypted:"]
2019/08/22 18:20:08 DEBUG : Using config file from "/data/data/com.termux/files/home/.config/rclone/rclone.conf"
2019/08/22 18:20:10 DEBUG : Adding path "vfs/forget" to remote control registry
2019/08/22 18:20:10 DEBUG : Adding path "vfs/refresh" to remote control registry
2019/08/22 18:20:10 DEBUG : Adding path "vfs/poll-interval" to remote control registry
2019/08/22 18:20:10 NOTICE: Encrypted drive 'Drive-crypted:': Serving HTTP on [::]:7879
2019/08/22 18:20:10 INFO  : DLNA server on :7879: Started SSDP on wlan0
2019/08/22 18:20:11 INFO  : /static/ContentDirectory.xml: 192.168.1.10:48660 GET 200  %!s(<nil>)
2019/08/22 18:20:11 INFO  : /static/ConnectionManager.xml: 192.168.1.10:48660 GET 200  %!s(<nil>)
2019/08/22 18:20:11 INFO  : /static/X_MS_MediaReceiverRegistrar.xml: 192.168.1.10:48660 GET 200  %!s(<nil>)
2019/08/22 18:20:22 INFO  : /static/ContentDirectory.xml: 192.168.1.8:43483 GET 200  %!s(<nil>)
2019/08/22 18:20:22 INFO  : /static/ConnectionManager.xml: 192.168.1.8:43483 GET 200  %!s(<nil>)
2019/08/22 18:20:22 INFO  : /static/X_MS_MediaReceiverRegistrar.xml: 192.168.1.8:43483 GET 200  %!s(<nil>)
2019/08/22 18:20:33 INFO  : /ctl: 192.168.1.8:43483 POST 200 "urn:schemas-upnp-org:service:ContentDirectory:1#Browse" %!s(<nil>)
2019/08/22 18:20:33 INFO  : /ctl: 192.168.1.8:43483 POST 200 "urn:schemas-upnp-org:service:ContentDirectory:1#Browse" %!s(<nil>)
2019/08/22 18:20:36 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.doc
2019/08/22 18:20:36 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.mp4
2019/08/22 18:20:36 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.rar
2019/08/22 18:20:36 INFO  : /ctl: 192.168.1.8:43483 POST 200 "urn:schemas-upnp-org:service:ContentDirectory:1#Browse" %!s(<nil>)
2019/08/22 18:20:36 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.doc
2019/08/22 18:20:36 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.mp4
2019/08/22 18:20:36 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.rar
2019/08/22 18:20:36 INFO  : /ctl: 192.168.1.8:43483 POST 200 "urn:schemas-upnp-org:service:ContentDirectory:1#Browse" %!s(<nil>)

</details>

#### Relevant discussion on the rclone forum
https://forum.rclone.org/t/error-using-rclone-serve-dlna-on-termux/11083/27

**Comment by mt025** [comment]
I just had a look at this and it seems that rclone uses the `mime` include for go. This references a mime type file that doesn't seem to exist in termux. `/etc/mime.types`  [https://golang.org/pkg/mime/#TypeByExtension](https://golang.org/pkg/mime/#TypeByExtension).

I created this file in `/data/data/com.termux/files/usr/etc/mime.types` but no luck.  However, with root if you create `/etc/mime.types` the same as in Debian it works as expected.

Seems to be more of an issue with go/mime or termux rather than rclone itself.

**Comment by ncw** [comment]
@mt025 that is a very interesting observation - thank you.

I had a look at the go source and yes you are right, unless termux patches the go source then it will be looking in `/etc/mime.types` I think...

Termux doesn't appear to patch go for that so potentially it could be if someone wanted to make a patch: https://github.com/termux/termux-packages/tree/master/packages/golang

However, looking at the serve dlna code, I see it is using the mime type database to read the mime type from the file name. However most backends know the mime type so we could ask the backend itself...

I implemented that here... That means it may start working for a drive backend (but won't for a local backend)

@serovar Can you give this a go?

https://beta.rclone.org/branch/v1.48.0-218-ge2a5e2c9-fix-3475-dlna-mime-beta/ (uploaded in 15-30 mins)

Or build it from the `fix-3475-dlna-mime` branch might be better!

If that doesn't work then we'll have to work on a patch for the go package in Termux.

**Comment by serovar** [comment]
I built the package from the `fix-3475-dlna-mime` branch, but unfortunately it has the same problem.

<details>
<summary>Here is the log</summary>
$rclone/rclone -vv serve dlna Drive-crypted:
2019/08/22 21:42:33 DEBUG : rclone: Version "v1.48.0-DEV" starting with parameters ["rclone/rclone" "-vv" "serve" "dlna" "Drive-crypted:"]
2019/08/22 21:42:33 DEBUG : Using config file from "/data/data/com.termux/files/home/.config/rclone/rclone.conf"
2019/08/22 21:42:34 DEBUG : Drive-notcrypted: Loaded invalid token from config file - ignoring
2019/08/22 21:42:34 DEBUG : Drive-notcrypted: Saved new token in config file
2019/08/22 21:42:36 DEBUG : Adding path "vfs/forget" to remote control registry
2019/08/22 21:42:36 DEBUG : Adding path "vfs/refresh" to remote control registry
2019/08/22 21:42:36 DEBUG : Adding path "vfs/poll-interval" to remote control registry
2019/08/22 21:42:36 NOTICE: Encrypted drive 'Drive-crypted:': Serving HTTP on [::]:7879
2019/08/22 21:42:36 INFO  : DLNA server on :7879: Started SSDP on wlan0
2019/08/22 21:42:37 INFO  : /static/ContentDirectory.xml: 192.168.1.10:50538 GET 200  %!s(<nil>)
2019/08/22 21:42:37 INFO  : /static/ConnectionManager.xml: 192.168.1.10:50538 GET 200  %!s(<nil>)
2019/08/22 21:42:37 INFO  : /static/X_MS_MediaReceiverRegistrar.xml: 192.168.1.10:50538 GET 200  %!s(<nil>)
2019/08/22 21:42:49 INFO  : /static/ContentDirectory.xml: 192.168.1.8:49967 GET 200  %!s(<nil>)
2019/08/22 21:42:49 INFO  : /static/ConnectionManager.xml: 192.168.1.8:49967 GET 200  %!s(<nil>)
2019/08/22 21:42:49 INFO  : /static/X_MS_MediaReceiverRegistrar.xml: 192.168.1.8:49967 GET 200  %!s(<nil>)
2019/08/22 21:42:50 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.doc
2019/08/22 21:42:50 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.mp4
2019/08/22 21:42:50 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.rar
2019/08/22 21:42:50 INFO  : /ctl: 192.168.1.8:49967 POST 200 "urn:schemas-upnp-org:service:ContentDirectory:1#Browse" %!s(<nil>)
2019/08/22 21:42:50 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.doc
2019/08/22 21:43:00 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.mp4
2019/08/22 21:43:00 DEBUG : DLNA server on :7879: unrecognized file type: Prova/advertisement_for_barcelona_s_tram_spain-p4gyc61hpf931.rar
2019/08/22 21:43:00 INFO  : /ctl: 192.168.1.8:49967 POST 200 "urn:schemas-upnp-org:service:ContentDirectory:1#Browse" %!s(<nil>)


</details>

**Comment by ncw** [comment]
> I built the package from the `fix-3475-dlna-mime` branch, but unfortunately it has the same problem.
> 
> Here is the log

Can you try serving something from an unencrypted drive remote?

**Comment by serovar** [comment]
It works!

**Comment by ncw** [comment]
> It works!

Can you compare with the original version - I suspect that won't work?

**Comment by serovar** [comment]
Exactly.

**Comment by ncw** [comment]
OK try this on the encrypted drive - hopefully it will work :-)

I added a very small number of mimetypes to go's internal store which should cover the most common audio and video files, hopefully that will work for you!

Find that on the same branch `fix-3475-dlna-mime` (`git pull` to update)

```go
		{"audio/flac", ".flac"},
		{"audio/mpeg", ".mpga,.mpega,.mp2,.mp3,.m4a"},
		{"audio/ogg", ".oga,.ogg,.opus,.spx"},
		{"audio/x-wav", ".wav"},
		{"image/tiff", ".tiff,.tif"},
		{"video/dv", ".dif,.dv"},
		{"video/fli", ".fli"},
		{"video/mpeg", ".mpeg,.mpg,.mpe"},
		{"video/MP2T", ".ts"},
		{"video/mp4", ".mp4"},
		{"video/quicktime", ".qt,.mov"},
		{"video/ogg", ".ogv"},
		{"video/webm", ".webm"},
		{"video/x-msvideo", ".avi"},
		{"video/x-matroska", ".mpv,.mkv"},
```

Or for a traditional beta:
https://beta.rclone.org/branch/v1.48.0-219-g40a9755a-fix-3475-dlna-mime-beta/ (uploaded in 15-30 mins)

**Comment by serovar** [comment]
The encrypted drive works perfectly, thank you very much for your incredible support!

**Comment by ncw** [comment]
> The encrypted drive works perfectly, thank you very much for your incredible support!

Great!  I wouldn't have got there without @mt025 so thanks due there too :-)

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.49
